kotsadm/pkg/handlers/airgap.go

```diff
@@ -89,49 +262,116 @@ func updateAppFromAirgap(w http.ResponseWriter, r *http.Request) {
+	if err := json.NewDecoder(r.Body).Decode(&createAppFromAirgapRequest); err != nil {
+		logger.Error(err)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
 	pendingApp, err := store.GetStore().GetPendingAirgapUploadApp()
 	if err != nil {
 		logger.Error(err)
-		w.WriteHeader(500)
+		w.WriteHeader(http.StatusInternalServerError)
 		return
 	}
 
 	var registryHost, namespace, username, password string
 	registryHost, username, password, err = getKurlRegistryCreds()
 	if err != nil {
 		logger.Error(err)
-		w.WriteHeader(500)
+		w.WriteHeader(http.StatusInternalServerError)
 		return
 	}
 
 	// if found kurl registry creds, use kurl registry
 	if registryHost != "" {
 		namespace = pendingApp.Slug
 	} else {
-		registryHost = r.FormValue("registryHost")
-		namespace = r.FormValue("namespace")
-		username = r.FormValue("username")
-		password = r.FormValue("password")
-	}
-
-	airgapBundle, _, err := r.FormFile("file")
-	if err != nil {
-		logger.Error(err)
-		w.WriteHeader(500)
-		return
+		registryHost = createAppFromAirgapRequest.RegistryHost
+		namespace = createAppFromAirgapRequest.Namespace
+		username = createAppFromAirgapRequest.Username
+		password = createAppFromAirgapRequest.Password
 	}
 
 	go func() {
-		defer airgapBundle.Close()
-		if err := airgap.CreateAppFromAirgap(pendingApp, airgapBundle, registryHost, namespace, username, password); err != nil {
-			logger.Error(err)
+		if err := airgap.CreateAppFromAirgap(pendingApp, airgapBundlePath, registryHost, namespace, username, password); err != nil {
+			logger.Error(errors.Wrap(err, "failed to create app from airgap bundle"))
+			return
+		}
+		// app created successfully, we can remove the airgap bundle
+		if err := os.RemoveAll(airgapBundlePath); err != nil {
+			logger.Error(errors.Wrap(err, "failed to remove airgap bundle after create"))
 		}
 	}()
 
 	createAppFromAirgapResponse := CreateAppFromAirgapResponse{}
 
-	JSON(w, 202, createAppFromAirgapResponse)
+	JSON(w, http.StatusOK, createAppFromAirgapResponse)
+}
+
+func getChunkName(uploadedFileName string, chunkNumber int64) string {
+	return fmt.Sprintf("%s_part_%d", uploadedFileName, chunkNumber)
+}
+
+func getAirgapBundlePath(uploadedFileIdentifier string) string {
+	return filepath.Join(os.TempDir(), fmt.Sprintf("%s.%s", uploadedFileIdentifier, "airgap"))
+}
+
+func isUploadComplete(uploadedFileName string, chunksDir string, totalChunks int64) (bool, error) {
+	isUploadComplete := true
+
+	var i int64
+	for i = 1; i <= totalChunks; i++ {
+		chunkName := getChunkName(uploadedFileName, i)
+		chunkPath := filepath.Join(chunksDir, chunkName)
+		_, err := os.Stat(chunkPath)
+		if os.IsNotExist(err) {
+			isUploadComplete = false
+		} else if err != nil {
+			return false, errors.Wrap(err, "failed to os state file")
+		}
+	}
+
+	return isUploadComplete, nil
+}
+
+func createAirgapBundleFromChunks(airgapBundlePath string, uploadedFileName string, chunksDir string, totalChunks int64) error {
```

Comment:
i'm not sure how long this will take for really big files, might be a better idea to write the chunks directly to the file and calculate offsets to determine if the file upload was complete.